### PR TITLE
perf: Enable property group optimizations for all new teams by default

### PR DESCRIPTION
**Pull request by tkaemming**
## Problem

New teams still use the slower path even though they don't have any data that needs backfilling.

## Changes

This changes the default value of `modifiers` so that all teams use the new path (see: #24152, #24171, #24381, #24897 for details.)

This could also be done with a feature flag but this has a couple advantages:

- we're already using `modifiers` in production for this, so change is minimal and avoids adding new layers of configuration (like what happened with the PoE rollout)
- allows hobby deployments to use the new functionality for new projects without any extra work beyond the typical "run migrations" step

It does have some disadvantages though, too:

- can't do double duty as a kill switch in an emergency
- it is less auditable who made changes and when
- includes a migration with a clumsy import back into application code

I couldn't really make up my mind on which approach was "best" so just defaulted to the least impact change since the likelihood that we'll need a kill switch seems minimal (we've been testing this for a while — and worst case we can just run an `UPDATE` to drop the modifier for all teams if we have to) and most if not all teams should be able to be migrated to this in the coming days. (After this is merged, I'm planning on updating all teams that were created after we started writing events with property groups to use this optimization, in fact.)

## Does this work well for both Cloud and self-hosted?

Yes: see above, this turns it on for new projects in self-hosted as well.

## How did you test this code?

See updated snapshots.

### CODE REVIEW

**Review by timgl — APPROVED** [review]

File: posthog/models/team/team.py
Comment (Twixes) [inline posthog/models/team/team.py:172]: I'm only worried about us  creating two ways of defining default modifiers here:
1. this, at team creation time,
2. and the existing `set_default_modifier_values`, at query time.
What do you think about setting `propertyGroupsMode` in `set_default_modifier_values()` based on `team.created_at`? One bonus would be enabling `propertyGroupsMode` retroactively, for team created a few days back let's say

**Comment by timgl** [comment]
What's stopping us from getting this in? Is it just the EU backfill?

**Comment by tkaemming** [comment]
> What's stopping us from getting this in? Is it just the EU backfill?

Technically nothing — this would only apply to teams created after this is merged, so it's not blocked by the backfill. The modifier would still need to be enabled manually for teams created before to this change.

I mostly wanted to give @Twixes's [comment](https://github.com/PostHog/posthog/pull/25010#discussion_r1771442330) more consideration and then got distracted with some of the index selection issues and other performance issues. I do think both approaches have downsides — we're either adding more methods to control which properties are enabled for a given team (adding a new flag), or adding more methods to control which _default_ properties are enabled for a given team (changing the default modifiers.) I'm not sure one of these is materially better or worse than the other.

The biggest distinction between the approaches, I think, is that changing the model default means this will be enabled for new hobby installs/teams, while using a feature flag means it won't. Does that matter? I also don't have a clear plan for how we're going to roll this out to _preexisting_ hobby teams, either: hobby doesn't have materialized columns by default (and the non-EE FOSS install doesn't seem to even have them at all), so maybe it's simpler just to _not_ make it the model default and use a feature flag, if that's the existing precedent?

Assuming no hiccups on EU over the next couple days, I think we should be able to roll this out to everybody (both clusters, cloud specifically) this week at the rate things have been backfilling. Whether that happens via a (new) feature flag, or via the existing `Team.modifiers` (as we've been doing) doesn't seem like a big issue to me, as long as new signups have this either via the feature flag options or the model default.

**Comment by Twixes** [comment]
I mostly think we should use `set_default_modifier_values` so that we can enable this retroactively for Clouds teams created in the past week or so. But that's not a blocker.
In either case I wouldn't consider hobby deploys, so long as we don't break things for them.

When data is fully backfilled, I would much recommend a quick `set_default_modifier_values()` change over a migration iterating over all the teams. We can then ignore on hobby deploys just to sidestep that problem:
```python3
if modifiers.propertyGroupsMode is None and is_cloud():
    modifiers.propertyGroupsMode = <enabled, or optimized - not sure how they differ>
```

**Pull request closed without merging**
